### PR TITLE
[Kernel] register punica functions as torch ops

### PR DESCRIPTION
**Pull request by bnellnm**
Register punica functions as torch ops to support dynamo and prevent graph breaks.

Broken off from https://github.com/vllm-project/vllm/pull/6917

cc @youkaichao, @SageMoore 

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/lora/ops/bgmv_expand.py [hunk 1]
```diff
@@ -108,6 +107,7 @@ def bgmv_expand(
         override_config (Optional[Dict[str, int]], optional): Defaults to None. 
             Triton grid config
     """
+    override_config: Optional[Dict[str, int]] = None
```
[the remark this answers is not in the document]
Comment (mgoin) [reply vllm/lora/ops/bgmv_expand.py:110]: Also the docstring should be updated

Comment (youkaichao) [inline vllm/lora/ops/bgmv_expand.py:110]: cc @jeejeelee what's the purpose of this code?

File: vllm/lora/ops/bgmv_expand.py [hunk 2]
```diff
@@ -167,3 +167,8 @@ def bgmv_expand(
         **config,
     )
     return
+
+
+bgmv_expand = torch.library.custom_op("lora::bgmv_expand",
```
Comment (youkaichao) [inline vllm/lora/ops/bgmv_expand.py:172]: do you need to register the fake implementation?

**Review by youkaichao — COMMENTED** [review]
@bnellnm add tests for these ops?

File: vllm/lora/ops/bgmv_expand.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (bnellnm) [reply vllm/lora/ops/bgmv_expand.py:110]: I forgot to clean this up.  Pytorch can't deal with `Dict` and this argument wasn't being used anywhere.

**Comment by bnellnm** [comment]
> @bnellnm add tests for these ops?

There's already tests for these.

File: vllm/lora/ops/bgmv_expand.py (hunk 2 above)
Comment (bnellnm) [reply vllm/lora/ops/bgmv_expand.py:172]: dynamo never complained about missing fake implementations.  although it seems like they should have them.

**Review by mgoin — APPROVED** [review]
LGTM since nothing was using override_config before this PR

File: vllm/lora/ops/bgmv_expand.py (hunk 1 above)
Comment (jeejeelee) [reply vllm/lora/ops/bgmv_expand.py]: This is the configuration passed when tuning kernel. We currently don't support tuning kernel for punica,  if tuning is implemented later, we can consider adding this parameter then.

File: vllm/lora/ops/bgmv_expand_slice.py [hunk 3]
```diff
@@ -180,3 +171,8 @@ def bgmv_expand_slice(
         **config,
     )
     return
+
+
+bgmv_expand_slice = torch.library.custom_op("lora::bgmv_expand_slice",
```
Comment (gshtras) [inline vllm/lora/ops/bgmv_expand_slice.py:176]: custom_op is torch 2.4+
Probably worth adding something similar to https://github.com/vllm-project/vllm/pull/7886/files#diff-dab7693bd00a09e22d39aee684a7e419aa358a47c4bd20df33d44f5adf60d304R1114-R1115 in punica.py:16 around the import

**Pull request merged**